商品出品機能

swatanabe1122 opened the pull request:
# What 
商品出品機能を実装
# Why
商品の情報が保存できるようにするため。

ログイン状態の場合は、商品出品ページへ遷移できる動画
https://gyazo.com/0e71bb0f82a11a98bbbbe52fc08e054a
 価格が入力されると同時に、販売手数料と販売利益が表示される動画（計算式にMath.floorを適用後）
https://gyazo.com/6d925fccf0e757a1250efeb0c44f51ce
 必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画
https://gyazo.com/c85b6f56899c00ac139dda80d4063b4d
 入力に問題がある状態で「出品する」ボタンが押された場合、情報は保存されず、出品ページに戻りエラーメッセージが表示される動画
https://gyazo.com/00de64241bdd8654dc3c2962568ece72
 ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/2df6de336ea966bea489e3568a684411
 テスト結果の画像（バリデーション修正後）
https://gyazo.com/fbeef01c37371e9dfeff62bbf4224b23

swatanabe1122 committed:
モデルとデータベース作成

Active Storageの導入、 ActiveHashを導入する、バリデーションを設定する 商品出品ページを表示できるようにする

swatanabe1122 committed:
カテゴリーの情報を実装

swatanabe1122 committed:
ビューファイルのhogeを正しくなるよう修正

swatanabe1122 committed:
必要なJSファイルを作成

JavaScriptで作成し、ブラウザ上に描画できるようにコードを記入。

swatanabe1122 committed:
バリデーションとテストコードの設定

swatanabe1122 committed:
エラーハンドリング画面の修正

minachanchan commented on app/controllers/items_controller.rb:
```diff
@@ -1,7 +1,34 @@
   # ユーザーがログインしていなければ、そのユーザーをログイン画面に遷移させる
   before_action :authenticate_user!, except: [:index, :show]
+
+  before_action :set_item, only: [:show, :edit, :update, :destroy]
```
```ruby
before_action :set_item, only: [:show, :edit, :update, :destroy]

  def set_item
    @item = Item.find(params[:id])
  end
```
5行目並びに26-28行目の記述は今回の実装では使用していない記述のようです。
該当機能実装時に差分として適切に出力できるよう、今は一度コメントアウトしておきましょう！

minachanchan commented on app/controllers/items_controller.rb:
```diff
@@ -1,7 +1,34 @@
+  before_action :set_item, only: [:show, :edit, :update, :destroy]
+
   def index
+     @items = Item.order("created_at DESC")
```
```ruby
@items = Item.order("created_at DESC")
```
こちらも商品一覧機能実装時に差分として出力できるようコメントアウトしましょう！

minachanchan commented on app/javascript/item_price.js:
```diff
@@ -0,0 +1,11 @@
+  const ProfitDom = document.getElementById("profit");
+  
+  addTaxDom.innerHTML = Math.floor(inputValue * 0.1);
+  ProfitDom.innerHTML = inputValue - addTaxDom.innerHTML;
```
```ruby
ProfitDom.innerHTML = inputValue - addTaxDom.innerHTML;
```
こちらも計算式にMath.floorを適用しましょう！

minachanchan commented on app/models/item.rb:
```diff
@@ -0,0 +1,22 @@
+  belongs_to :shipping_cost
+  belongs_to :user
+  has_one_attached :image
+  has_one :order
```
```ruby
has_one :order
```
こちらも今回の実装内容に含まれませんので、コメントアウトしておきましょう！

minachanchan commented on app/models/item.rb:
```diff
@@ -0,0 +1,22 @@
+class Item < ApplicationRecord
+  extend ActiveHash::Associations::ActiveRecordExtensions
+  belongs_to :item_status
+  belongs_to :shipping_cost
```
```ruby
  belongs_to :item_status
  belongs_to :shipping_cost
```
active_hashで実装した5つの全てのクラスに対するアソシエーションを記載しましょう！

minachanchan commented on app/views/items/index.html.erb:
```diff
@@ -123,66 +123,66 @@
-        <%= link_to "#" do %>
-        <div class='item-img-content'>
-          <%= image_tag "item-sample.png", class: "item-img" %>
+      <!-- 実際の商品データの表示 -->
```
こちらのビューファイルに実装された内容は全て元に戻すか、エラーの発生しないようコメントアウトしましょう！
こちらは商品一覧機能の実装内容のためです。

minachanchan commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,128 @@
+        @item.price = 10_000_000
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price must be less than or equal to 9999999')
+      end
```
商品価格に関しては、以下を確認するテストコードをそれぞれ実装しましょう！

```
・半角数字以外の値が含まれている場合は保存できないこと
```

minachanchan reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。
キャプチャのご提示もお礼申し上げます！

コメントいたしましたので、ご確認頂けますと幸いでございます。


実装完了まであと少し、一緒に頑張りましょう！
修正の後、再度レビューのご依頼頂けますようお願い申し上げます。

swatanabe1122 committed:
商品一覧機能実装時に差分として出力できるようコメントアウト

swatanabe1122 committed:
半角数字以外の値が含まれている場合は保存できないこと

商品一覧機能の実装内容のための記述を
エラーの発生しないようコメントアウト。
active_hashで実装した5つの全てのクラスに対するアソシエーションを記載後バリデーションも修正

swatanabe1122 commented:
価格が入力されると同時に、販売手数料と販売利益が表示される動画（計算式にMath.floorを適用後）
https://gyazo.com/6d925fccf0e757a1250efeb0c44f51ce
バリデーションを修正後の、テスト結果の画像https://gyazo.com/fbeef01c37371e9dfeff62bbf4224b23

TakashiKozuma commented:
追加のレビュー依頼ありがとうございます！
修正点も含め、確認が完了しました。
LGTMです！！

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に次の実装のためのブランチを切ってから次にお進み下さい。
